core/src/repair/repair_service.rs

```diff
@@ -488,220 +476,306 @@ impl RepairService {
+        root_bank: Arc<Bank>,
+        repair_weight: &mut RepairWeight,
+        popular_pruned_forks_requests: &mut HashSet<Slot>,
+        popular_pruned_forks_sender: &PopularPrunedForksSender,
+        repair_metrics: &mut RepairMetrics,
+    ) {
+        let mut handle_popular_pruned_forks = Measure::start("handle_popular_pruned_forks");
+        let mut popular_pruned_forks = repair_weight
+            .get_popular_pruned_forks(root_bank.epoch_stakes_map(), root_bank.epoch_schedule());
+        // Check if we've already sent a request along this pruned fork
+        popular_pruned_forks.retain(|slot| {
+            if popular_pruned_forks_requests
+                .iter()
+                .any(|prev_req_slot| repair_weight.same_tree(*slot, *prev_req_slot))
+            {
+                false
+            } else {
+                popular_pruned_forks_requests.insert(*slot);
+                true
+            }
+        });
+        if !popular_pruned_forks.is_empty() {
+            warn!(
+                "Notifying repair of popular pruned forks {:?}",
+                popular_pruned_forks
+            );
+            popular_pruned_forks_sender
+                .send(popular_pruned_forks)
+                .unwrap_or_else(|err| error!("failed to send popular pruned forks {err}"));
+        }
+        handle_popular_pruned_forks.stop();
+        repair_metrics.timing.handle_popular_pruned_forks = handle_popular_pruned_forks.as_us();
+    }
+
+    fn build_and_send_repair_batch(
+        serve_repair: &mut ServeRepair,
+        peers_cache: &mut LruCache<u64, RepairPeers>,
+        repair_request_quic_sender: &AsyncSender<(SocketAddr, Bytes)>,
+        repairs: Vec<ShredRepairType>,
+        repair_info: &RepairInfo,
+        outstanding_requests: &RwLock<OutstandingShredRepairs>,
+        repair_socket: &UdpSocket,
+        repair_protocol: Protocol,
+        repair_metrics: &mut RepairMetrics,
+    ) {
+        let mut build_repairs_batch_elapsed = Measure::start("build_repairs_batch_elapsed");
+        let batch: Vec<(Vec<u8>, SocketAddr)> = {
+            let mut outstanding_requests = outstanding_requests.write().unwrap();
+            repairs
+                .into_iter()
+                .filter_map(|repair_request| {
+                    let (to, req) = serve_repair
+                        .repair_request(
+                            &repair_info.cluster_slots,
+                            repair_request,
+                            peers_cache,
+                            &mut repair_metrics.stats,
+                            &repair_info.repair_validators,
+                            &mut outstanding_requests,
+                            &repair_info.cluster_info.keypair(),
+                            repair_request_quic_sender,
+                            repair_protocol,
+                        )
+                        .ok()??;
+                    Some((req, to))
+                })
+                .collect()
```

Comment:
This `.collect()` looks a little suspicious. I guess we need it since `serve_repair.repair_request()` is fallible, but I wonder if it is possible to refactor to avoid it.

Don't need to address in this PR, just mentioning it since it showed up in the diff

Reply:
yeah, good call